File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix:
            return False
        
        m = len(matrix)
        n = len(matrix[0])
        left = 0
        right = m * n - 1

        while left <= right:
            mid = (left + right) // 2

            row = mid // n
            column = mid % n
            print(mid, row, column)
            if matrix[row][column] == target:
                return True
            elif matrix[row][column] < target:
                left = mid + 1
            else:
                right = mid - 1
            
        return False

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int numRows = matrix.length;
        int numCols = matrix[0].length;
        int totalElements = numRows * numCols;

        int left = 0, right = totalElements - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            int midRow = mid / numCols;
            int midCol = mid % numCols;
            int midELe = matrix[midRow][midCol];
            if (midELe == target) {
                return true;
            }

            if (midELe > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return false;
    }
}